### PR TITLE
On-Chain-Governance

### PR DESCRIPTION
**Pull request by bvbfan**
/kind feature

Fixes #576 #869

### CODE REVIEW

File: src/masternodes/proposals.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+// Distributed under the MIT software license, see the accompanying
+// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
+
+#include <masternodes/proposals.h>
+
+std::string CPropTypeToString(const CPropType status)
+{
+    switch(status) {
+    case CPropType::CommunityFundRequest:   return "CommunityFundRequest";
+    case CPropType::BlockRewardRellocation: return "BlockRewardRellocation";
```
Comment (Bushstar) [inline src/masternodes/proposals.cpp:11]: Very minor issue, Rellocation is a typo, it should be Reallocation.

File: src/validation.cpp [hunk 2]
```diff
@@ -2811,6 +2828,88 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
...
+                        case CPropType::BlockRewardRellocation:
+                            majorityThreshold = chainparams.GetConsensus().props.brp.majorityThreshold;
+                            break;
+                        case CPropType::VoteOfConfidence:
+                            majorityThreshold = chainparams.GetConsensus().props.voc.majorityThreshold;
+                            break;
+                        default:
+                            assert(false);
+                    }
+                    auto approved = lround(voteYes * 10000.f / voters) >= majorityThreshold;
```
Comment (Bushstar) [inline src/validation.cpp:2886]: Pink paper says simple majority required for CFP, so yes vote should be more than 50%, not equal to it. 50% would be an equal split of yes and no and should not side with a yes vote under this circumstance. The pink paper says the same for VoC and block reallocation, that it should be more than 66.67% and not equal, though in this case more than or equal is less disagreeable.

File: src/masternodes/rpc_proposals.cpp [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+
+/*
+ *  Issued by: any
+*/
+UniValue createcfp(const JSONRPCRequest& request)
+{
+    auto pwallet = GetWallet(request);
+
+    RPCHelpMan{"createcfp",
+               "\nCreates a Cummunity Fund Request" +
```
Comment (Bushstar) [inline src/masternodes/rpc_proposals.cpp:39]: Typo, should be "Community". Also createcfp should be createcfr as it is a community fund request. Pinkpaper refers to the call as creategovcfr so uses cfr as well. I think we can take some liberty with the actual call name, but if we use cfp then the help should be renamed Community Fund Proposal.

Comment (bvbfan) [reply src/masternodes/rpc_proposals.cpp:39]: Pinkpaper should be updated to show correct nick name `createcfp`, i will update the help, it should be `Community Fund Proposal`.

**Review by Bushstar — APPROVED** [review]

File: src/validation.cpp [hunk 4]
```diff
@@ -3291,6 +3311,91 @@ void CChainState::ProcessGovEvents(const CBlockIndex* pindex, CCustomCSView& cac
...
+        return;
+    }
+
+    if (pindex->nHeight == chainparams.GetConsensus().GreatWorldHeight) {
+        auto balance = cache.GetBalance(chainparams.GetConsensus().foundationShareScript, DCT_ID{0});
+        if (balance.nValue > 0) {
+            cache.SubBalance(chainparams.GetConsensus().foundationShareScript, balance);
+            cache.AddCommunityBalance(CommunityAccountType::CommunityDevFunds, balance.nValue);
+        }
+    }
```
Comment (bvbfan) [inline src/validation.cpp:3325]: I will gonna to remove that check, whether foundation receives tokens they will be transferred to `Community Dev Funds`

Comment (prasannavl) [reply src/validation.cpp:3325]: yup, sounds good.

File: src/validation.cpp [hunk 5]
```diff
@@ -2585,8 +2602,8 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
         {
             if (tx.vout[j].scriptPubKey == Params().GetConsensus().burnAddress)
             {
-                pburnHistoryDB->WriteAccountHistory({Params().GetConsensus().burnAddress, static_cast<uint32_t>(pindex->nHeight), i},
-                                            {block.vtx[i]->GetHash(), static_cast<uint8_t>(CustomTxType::None), {{DCT_ID{0}, tx.vout[j].nValue}}});
+                pburnHistoryDB->WriteAccountHistory({tx.vout[j].scriptPubKey, static_cast<uint32_t>(pindex->nHeight), i},
+                                                    {block.vtx[i]->GetHash(), static_cast<uint8_t>(CustomTxType::None), {{DCT_ID{0}, tx.vout[j].nValue}}});
```
Comment (chanakasameera) [inline src/validation.cpp:2606]: Will this cause the issue "`AccountHistory.blockHash` returned from `getaccounthistory` becomes undefined" ?
Related jellyfish test: https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-api-core/__tests__/category/account/getAccountHistory.test.ts#L46

**Review by chanakasameera — CHANGES_REQUESTED** [review]

**Review by chanakasameera — CHANGES_REQUESTED** [review]
Looks like `getmasternodeblocks` does not work as expected, it should return the number of blocks minted by a masternode ?, but now it returns a higher value.
Related jellyfish test: https://github.com/DeFiCh/jellyfish/blob/main/packages/jellyfish-api-core/__tests__/category/masternode/getMasternodeBlocks.test.ts#L178

File: src/validation.cpp (hunk 5 above)
Comment (bvbfan) [reply src/validation.cpp:2606]: No, it shouldn't

**Comment by bvbfan** [comment]
Unrelated to current PR

**Pull request merged**